PermissionVault/contracts/PermissionVault.sol

```diff
@@ -0,0 +1,348 @@
+
+    token.safeTransferFrom(msg.sender, address(this), amount);
+    emit DepositERC20(msg.sender, token, amount);
+  }
+
+  /**
+   * @notice A controller can use this function to withdraw ERC20 tokens from the contract.
+   * @param to the address that will receive the withdrawn tokens
+   * @param token the ERC20 token contract address
+   * @param amount the amount of tokens to withdraw
+   */
+  function withdrawERC20(
+    address to,
+    IERC20 token,
+    uint256 amount
+  ) external onlyController {
+    if (amount > token.balanceOf(address(this))) {
+      revert NotEnoughBalance();
+    }
+
+    token.safeTransfer(to, amount);
+    emit WithdrawERC20(msg.sender, to, token, amount);
+  }
+
+  /**
+   * @notice A controller can use this function to deposit ERC721 tokens into the contract.
+   * @param token the ERC721 token contract address
+   * @param id the token ID to deposit
+   */
+  function depositERC721(IERC721 token, uint256 id) external {
+    if (token.ownerOf(id) != msg.sender) {
+      revert NotOwnerOf();
+    }
+    token.safeTransferFrom(msg.sender, address(this), id);
+    emit DepositERC721(msg.sender, token, id);
+  }
+
+  /**
+   * @notice A controller can use this function to withdraw ERC721 tokens from the contract.
+   * @param to the address that will receive the withdrawn token
+   * @param token the ERC721 token contract address
+   * @param id the token ID to withdraw
+   */
+  function withdrawERC721(
+    address to,
+    IERC721 token,
+    uint256 id
+  ) external onlyController {
+    if (token.ownerOf(id) != address(this)) {
+      revert NotExistToken();
+    }
+
+    token.safeTransferFrom(address(this), to, id);
+    emit WithdrawERC721(msg.sender, to, token, id);
+  }
+
+  /**
+   * @notice A controller can use this function to deposit ERC1155 tokens into the contract.
+   * @param token the ERC1155 token contract address
+   * @param id the token ID to deposit
+   * @param amount the amount of tokens to deposit
+   */
+  function depositERC1155(IERC1155 token, uint256 id, uint256 amount) external {
+    if (amount == 0) {
+      revert ZeroAmount();
+    }
+    if (token.balanceOf(msg.sender, id) < amount) {
+      revert NotEnoughBalance();
+    }
+
+    token.safeTransferFrom(msg.sender, address(this), id, amount, "");
+    emit DepositERC1155(msg.sender, token, id, amount);
+  }
+
+  /**
+   * @notice A controller can use this function to withdraw ERC1155 tokens from the contract.
+   * @param to the address that will receive the withdrawn token
+   * @param token the ERC1155 token contract address
+   * @param id the token ID to withdraw
+   * @param amount the amount of tokens to withdraw
+   */
+  function withdrawERC1155(
+    address to,
+    IERC1155 token,
+    uint256 id,
+    uint256 amount
+  ) external onlyController {
+    if (amount > token.balanceOf(address(this), id)) {
+      revert NotEnoughBalance();
+    }
+
+    token.safeTransferFrom(address(this), to, id, amount, "");
```

Comment:
Please check allowance before safeTransferFrom

Reply:
don't need it